File: python/mirage/commands/cli/builtin/git/inspect.py
import asyncio
import re
from io import BytesIO

from dulwich.config import ConfigFile
from dulwich.repo import BaseRepo

from mirage.commands.cli.builtin.git.errors import GitError
from mirage.commands.cli.builtin.git.history import (LogFlags, parse_flags,
                                                     ref_commits, select)
from mirage.commands.cli.builtin.git.io import read_file
from mirage.commands.cli.builtin.git.revparse import split_revisions
from mirage.commands.cli.builtin.git.session import opened
from mirage.commands.cli.builtin.git.util import (check_operands, escaped,
                                                  fatal, start_point)
from mirage.commands.cli.types import CLIDoors, CLIInvocation
from mirage.commands.spec.flag_view import FlagView
from mirage.io.types import ByteSource, IOResult
from mirage.utils.posix import translate_classes
from mirage.version import __version__


async def repo_config(inv: CLIInvocation[None], fl: FlagView) -> ConfigFile:
    doors = inv.doors or CLIDoors()
    _, location = await opened(fl, doors)
    assert doors.dispatch is not None
    return ConfigFile.from_file(
        BytesIO(await read_file(doors.dispatch,
                                f"{location.commondir}/config")))


async def remote(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    try:
        check_operands(inv.texts, marked=escaped(inv.argv))
        cfg = await repo_config(inv, fl)
        lines = []
        for section in sorted(cfg.sections()):
            if len(section) != 2 or section[0] != b"remote":
                continue
            name = section[1].decode()
            if not fl.as_bool("verbose"):
                lines.append(name)
                continue
            values = list(cfg.items(section))
            urls = [v.decode() for k, v in values if k == b"url"]
            push = [v.decode() for k, v in values if k == b"pushurl"] or urls
            if urls:
                lines.append(f"{name}\t{urls[0]} (fetch)")
            lines.extend(f"{name}\t{url} (push)" for url in push)
        return ("".join(f"{line}\n" for line in lines).encode(), IOResult())
    except GitError as exc:
        return fatal(exc)


async def config(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    try:
        cfg = await repo_config(inv, fl)
        listing = fl.as_bool("list")
        regexp = fl.as_bool("get_regexp")
        origin = fl.as_bool("show_origin")
        if not listing and not inv.texts:
            return None, IOResult(exit_code=129,
                                  stderr=b"error: wrong number of arguments\n")
        key = inv.texts[0] if inv.texts else ""
        try:
            pattern = re.compile(translate_classes(
                config_key(key))) if regexp else None
        except re.error:
            return None, IOResult(
                exit_code=6,
                stderr=f"error: invalid key pattern: {key}\n".encode())
        values = []
        for section in cfg.sections():
            for name, value in cfg.items(section):
                full = b".".join((*section, name.lower())).decode()
                if listing or (pattern.search(full)
                               if pattern else full == config_key(key)):
                    values.append((full, value.decode()))
        if not listing and not regexp:
            values = values[-1:]
        prefix = ""
        if origin:
            _, location = await opened(fl, inv.doors or CLIDoors())
            source = f"{location.commondir}/config"
            ordinary = location.commondir == location.worktree + "/.git"
            if ordinary and start_point(fl) == location.worktree:
                source = ".git/config"
            prefix = f"file:{source}\t"
        lines = [
            prefix + (name +
                      ("=" if listing else " ") if listing or regexp else "") +
            value + "\n" for name, value in values
        ]
        return "".join(lines).encode(), IOResult(
            exit_code=0 if values or listing else 1)

    except GitError as exc:
        return fatal(exc)


def _show_refs(repo: BaseRepo, patterns: tuple[str, ...]) -> bytes:
    return "".join(f"{repo.refs[ref].decode()} {ref.decode()}\n"
                   for ref in sorted(repo.refs.allkeys())
                   if ref.startswith(b"refs/") and (not patterns or any(
                       ref.decode() == p or ref.decode().endswith('/' + p)
                       for p in patterns))).encode()


async def show_ref(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    try:
        repo, _ = await opened(FlagView(inv.flags), inv.doors or CLIDoors())
        out = await asyncio.to_thread(_show_refs, repo, tuple(inv.texts))
        return out, IOResult(exit_code=0 if out else 1)
    except GitError as exc:
        return fatal(exc)


def _revisions(repo: BaseRepo, revisions: tuple[str, ...],
               flags: LogFlags) -> list[bytes]:
    starts, hidden = split_revisions(repo, revisions)
    if flags.all_refs:
        starts[:0] = ref_commits(repo)
    return [commit.id for commit in select(repo, starts, flags, tuple(hidden))]


async def rev_list(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    fl = FlagView(inv.flags)
    try:
        check_operands(inv.texts, marked=escaped(inv.argv))
        repo, _ = await opened(fl, inv.doors or CLIDoors())
        flags = parse_flags(fl)
        if not inv.texts and not flags.all_refs:
            return None, IOResult(
                exit_code=129,
                stderr=b"usage: git rev-list [<options>] <commit>...\n")
        commits = await asyncio.to_thread(_revisions, repo, tuple(inv.texts),
                                          flags)
        return (f"{len(commits)}\n".encode() if fl.as_bool("count") else
                b"".join(oid + b"\n" for oid in commits)), IOResult()
    except GitError as exc:
        return fatal(exc)


async def version(
        inv: CLIInvocation[None]) -> tuple[ByteSource | None, IOResult]:
    return f"git version {__version__} (Mirage)\n".encode(), IOResult()


def config_key(key: str) -> str:
    parts = key.split('.')
    parts[0] = parts[0].lower()
    parts[-1] = parts[-1].lower()
    return '.'.join(parts)

File: typescript/packages/core/src/commands/cli/builtin/git/inspect.ts
import git from 'isomorphic-git'
import { VERSION } from '../../../../version.ts'

import { translateClasses } from '../../../../utils/posix.ts'
import { compareCodePoints } from '../../../../utils/sort.ts'
import { IOResult } from '../../../../io/types.ts'
import type { CommandFnResult } from '../../../config.ts'
import { FlagView } from '../../../spec/flag_view.ts'
import type { CLIInvocation } from '../../types.ts'
import { GitError } from './errors.ts'
import { parseFlags, refCommits, select } from './history.ts'
import { opened, repoArgs } from './repo.ts'
import { loadRefs } from './refs.ts'
import { readFile } from './io.ts'
import { splitRevisions } from './revparse.ts'
import { checkOperands, escaped, fatal, startPoint } from './util.ts'

const ENC = new TextEncoder()

export async function remote(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  try {
    checkOperands([...inv.texts], undefined, escaped(inv.argv))
    const repo = await opened(fl, inv.doors ?? {})
    const rows = (await git.listRemotes(repoArgs(repo))).sort((a, b) =>
      compareCodePoints(a.remote, b.remote),
    )
    const lines: string[] = []
    for (const row of rows) {
      if (!fl.asBool('verbose')) lines.push(row.remote)
      else {
        const urls = (await git.getConfigAll({
          ...repoArgs(repo),
          path: `remote.${row.remote}.url`,
        })) as string[]
        const push = (await git.getConfigAll({
          ...repoArgs(repo),
          path: `remote.${row.remote}.pushurl`,
        })) as string[]
        if (urls[0]) lines.push(`${row.remote}\t${urls[0]} (fetch)`)
        for (const url of push.length ? push : urls) lines.push(`${row.remote}\t${url} (push)`)
      }
    }
    return [ENC.encode(lines.length ? `${lines.join('\n')}\n` : ''), new IOResult()]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

export async function config(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  try {
    const repo = await opened(fl, inv.doors ?? {})
    const listing = fl.asBool('list'),
      regexp = fl.asBool('get_regexp'),
      origin = fl.asBool('show_origin')
    if (!listing && !inv.texts.length)
      return [
        null,
        new IOResult({ exitCode: 129, stderr: ENC.encode('error: wrong number of arguments\n') }),
      ]
    const key = inv.texts[0] ?? ''
    let pattern: RegExp | null
    try {
      pattern = regexp ? new RegExp(translateClasses(configKey(key))) : null
    } catch (err) {
      if (!(err instanceof SyntaxError)) throw err
      return [
        null,
        new IOResult({ exitCode: 6, stderr: ENC.encode(`error: invalid key pattern: ${key}\n`) }),
      ]
    }
    const text = new TextDecoder().decode(
      await readFile(repo.dispatch, `${repo.location.commondir}/config`),
    )
    let section = ''
    const keys: string[] = []
    for (const line of text.split('\n')) {
      const header = /^\s*\[([\w.-]+)(?:\s+"((?:[^"\\]|\\.)*)")?\]/.exec(line)
      if (header) {
        section =
          (header[1] ?? '').toLowerCase() +
          (header[2] === undefined ? '' : '.' + header[2].replace(/\\(.)/g, '$1'))
        continue
      }
      const entry = /^\s*([\w-]+)\s*(?:=|$)/.exec(line)
      if (entry && section) keys.push(section + '.' + (entry[1] ?? '').toLowerCase())
    }
    const values: [string, string][] = [],
      seen = new Map<string, number>(),
      cache = new Map<string, (string | boolean)[]>()
    for (const name of keys) {
      const at = seen.get(name) ?? 0
      seen.set(name, at + 1)
      if (!(listing || (pattern ? pattern.test(name) : name === configKey(key)))) continue
      let held = cache.get(name)
      if (!held) {
        held = await git.getConfigAll({ ...repoArgs(repo), path: name })
        cache.set(name, held)
      }
      values.push([name, String(held[at] ?? '')])
    }
    const chosen = listing || regexp ? values : values.slice(-1)
    const source =
      repo.location.commondir === repo.location.worktree + '/.git' &&
      startPoint(fl) === repo.location.worktree
        ? '.git/config'
        : repo.location.commondir + '/config'
    const prefix = origin ? `file:${source}\t` : ''
    const out = chosen
      .map(
        ([name, value]) =>
          prefix + (listing || regexp ? name + (listing ? '=' : ' ') : '') + value + '\n',
      )
      .join('')
    return [ENC.encode(out), new IOResult({ exitCode: chosen.length || listing ? 0 : 1 })]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

export async function showRef(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  try {
    const repo = await opened(fl, inv.doors ?? {})
    const refs = await loadRefs(repo.dispatch, repo.location.gitdir, repo.location.commondir)
    const lines: string[] = []
    for (const ref of [...refs.keys()]
      .filter((r) => r.startsWith('refs/'))
      .sort(compareCodePoints)) {
      if (
        inv.texts.length &&
        !inv.texts.some((pattern) => ref === pattern || ref.endsWith(`/${pattern}`))
      )
        continue
      const oid = await git.resolveRef({ ...repoArgs(repo), ref })
      lines.push(`${oid} ${ref}`)
    }
    return [
      ENC.encode(lines.length ? `${lines.join('\n')}\n` : ''),
      new IOResult({ exitCode: lines.length ? 0 : 1 }),
    ]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

export async function revList(inv: CLIInvocation): Promise<CommandFnResult> {
  const fl = new FlagView(inv.flags)
  try {
    checkOperands([...inv.texts], undefined, escaped(inv.argv))
    const repo = await opened(fl, inv.doors ?? {})
    const flags = parseFlags(fl)
    if (!inv.texts.length && !flags.allRefs)
      return [
        null,
        new IOResult({
          exitCode: 129,
          stderr: ENC.encode('usage: git rev-list [<options>] <commit>...\n'),
        }),
      ]
    const [shown, hidden] = await splitRevisions(repo, inv.texts)
    const starts = flags.allRefs ? [...(await refCommits(repo)), ...shown] : shown
    const commits = await select(repo, starts, flags, hidden)
    const out = fl.asBool('count')
      ? `${String(commits.length)}\n`
      : commits.map((c) => `${c.oid}\n`).join('')
    return [ENC.encode(out), new IOResult()]
  } catch (err) {
    if (err instanceof GitError) return fatal(err)
    throw err
  }
}

export function version(_inv: CLIInvocation): CommandFnResult {
  return [ENC.encode(`git version ${VERSION} (Mirage)\n`), new IOResult()]
}

function configKey(key: string): string {
  const parts = key.split('.')
  return parts
    .map((part, i) => (i === 0 || i === parts.length - 1 ? part.toLowerCase() : part))
    .join('.')
}
